From 89a18aaf344e371c761b0c4050c449d2adc39fdf Mon Sep 17 00:00:00 2001
From: nchernikov <nchernikov@maxlinear.com>
Date: Sun, 24 Aug 2025 12:50:55 +0300
Subject: [PATCH] WLANRTSYS-93436 Allow new bg cac on new channel if one is
 already going

Issue:
ZWDFS channel switch command for new channel is rejected if there is already ongoing bg cac

Fix:
Allow new bg cac for new channel and BW if requested via ctrl iface

Fix up to: WLANRTSYS-64810 ZWDFS changes for wave700 in iwlwav-hostap-ng.

Signed-off-by: nchernikov <nchernikov@maxlinear.com>
---
 hostapd/ctrl_iface.c | 16 ++++++++++++--
 src/ap/mxl_zwdfs.c   | 51 +++++++++++++++++++++++++++++++++++++++++++-
 src/ap/mxl_zwdfs.h   |  1 +
 3 files changed, 65 insertions(+), 3 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index cef64e843..31ba01e7e 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -3145,6 +3145,7 @@ static int hostapd_ctrl_iface_chan_switch(struct hostapd_iface *iface,
 		settings.freq_params.channel = chan;
 #ifdef CONFIG_VENDOR_MXL
 		int bgcac_possible;
+		int start_bg_cac;
 
 		bgcac_possible = mxl_hostapd_background_cac_possible(iface, &settings, switch_type_zwdfs);
 
@@ -3152,8 +3153,19 @@ static int hostapd_ctrl_iface_chan_switch(struct hostapd_iface *iface,
 			return bgcac_possible;
 		}
 
-		if (bgcac_possible)
-			return mxl_hostapd_start_background_cac(iface, &settings);
+		if (bgcac_possible) {
+			/*
+				Prepare new bg cac by doing:
+				- No BG CAC in progress - do nothing and start
+				- BG CAC in progress and request is on new channel - stop CAC and start
+				- Report fail if was unable to disable ZWDFS antenna and stop CAC if needed
+			*/
+			start_bg_cac = mxl_hostapd_prepare_new_background_cac(iface, &settings);
+			if (start_bg_cac > 0)
+				return mxl_hostapd_start_background_cac(iface, &settings);
+			else
+				return -1;
+		}
 #endif /* CONFIG_VENDOR_MXL */
 		wpa_printf(MSG_DEBUG,
 			   "DFS/CAC to (channel=%u, freq=%d, sec_channel_offset=%d, bw=%d, center_freq1=%d)",
diff --git a/src/ap/mxl_zwdfs.c b/src/ap/mxl_zwdfs.c
index 519ee8c18..3f9f28d61 100644
--- a/src/ap/mxl_zwdfs.c
+++ b/src/ap/mxl_zwdfs.c
@@ -28,6 +28,55 @@ static bool mxl_hostapd_is_zwdfs_iface(struct hostapd_iface *iface)
 	return !!(iface->mxl_iface.drv_flags & MXL_WPA_DRIVER_FLAGS_BACKGROUND_CAC) || mxl_dfs_use_radar_background(iface);
 }
 
+/*
+	Function is used in ctrl iface CHAN_SWITCH command with ZWDFS switch type
+	Used to determine if ongoing BG CAC needs to be canceled in order to start
+	a new one on different channel
+
+	Returns:
+	 1 - New BG CAC can be started - either no ongoing CAC or requested on different channel and BW
+	 0 - New BG CAC cannot be started/restarted as ongoing CAC is already on same channel and BW
+	-1 - Error disabling ongoing BG CAC
+*/
+int mxl_hostapd_prepare_new_background_cac(struct hostapd_iface *iface, struct csa_settings *settings)
+{
+	struct hostapd_freq_params *new_freq = (struct hostapd_freq_params *)&settings->freq_params;
+	uint8_t new_centr_freq_seg0_idx = 0;
+	bool is_new_freq = false;
+
+	/* Nothing to do as there is no ongoing bg cac - allow new BG CAC */
+	if (!iface->radar_background.cac_started)
+		return 1;
+
+	if (settings->freq_params.bandwidth == 20) {
+		new_centr_freq_seg0_idx = new_freq->channel;
+#ifdef CONFIG_VENDOR_MXL_PROPRIETARY
+	} else {
+		int chan_offset = 0;
+		mxl_hostapd_get_chan_data_from_candidates_5g(new_freq->channel,
+							     mxl_get_oper_width(new_freq->bandwidth),
+							     &chan_offset, &new_centr_freq_seg0_idx);
+#endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
+	}
+	is_new_freq = (new_freq->channel != iface->radar_background.channel) ||
+				(new_freq->freq != iface->radar_background.freq) ||
+				(new_centr_freq_seg0_idx != iface->radar_background.centr_freq_seg0_idx);
+
+	/* CAC is already ongoing on same channel - dont start new one */
+	if (!is_new_freq) {
+		wpa_printf(MSG_WARNING, "ZWDFS: Background CAC already running on the same channel");
+		return 0;
+	}
+
+	/* Stop ongoing CAC to start new one on different channel */
+	if (mxl_hostapd_set_zwdfs_antenna(iface->bss[0], 0)) {
+		wpa_printf(MSG_ERROR, "ZWDFS: Failed to disable antenna on bg cac prepare");
+		return -1;
+	}
+
+	return 1;
+}
+
 int mxl_hostapd_background_cac_possible(struct hostapd_iface *iface, struct csa_settings *settings,
 					bool switch_type_zwdfs)
 {
@@ -55,7 +104,7 @@ int mxl_hostapd_background_cac_possible(struct hostapd_iface *iface, struct csa_
 		if (!mxl_dfs_use_radar_background(iface)) {
 			return -1;
 		}
-		if ((iface->radar_background.cac_started) || (iface->state < HAPD_IFACE_ACS)) {
+		if (iface->state < HAPD_IFACE_ACS) {
 			return -1;
 		}
 		res = mxl_zwdfs_check_chans_unavailable(iface, &settings->freq_params);
diff --git a/src/ap/mxl_zwdfs.h b/src/ap/mxl_zwdfs.h
index 4633b7014..e1c1a63f0 100644
--- a/src/ap/mxl_zwdfs.h
+++ b/src/ap/mxl_zwdfs.h
@@ -23,6 +23,7 @@ int mxl_hostapd_drv_get_zwdfs_antenna(struct hostapd_data *hapd, int *enabled);
 int mxl_hostapd_set_zwdfs_antenna(struct hostapd_data *hapd, int enable);
 bool mxl_hostapd_check_only_dfs_chans(struct hostapd_iface *iface);
 int mxl_hostapd_background_cac_possible(struct hostapd_iface *iface, struct csa_settings *settings, bool switch_type_zwdfs);
+int mxl_hostapd_prepare_new_background_cac(struct hostapd_iface *iface, struct csa_settings *settings);
 int mxl_hostapd_start_background_cac(struct hostapd_iface *iface, const struct csa_settings *settings);
 int mxl_hostapd_zwdfs_switch(struct hostapd_iface *iface, int cac_status, int freq, int chan_width, int cf1, int cf2);
 
-- 
2.43.0

